[CORE-13364] admin: Add a scale test for ListKafkaConnections

BenPope opened the pull request:
* Add a scale test of 10000 consumers across many groups.
* Add a benchmark:

```
ListKafkaConnectionsTest.CollectProto100k                        5    13.781ms    76.531us    13.705ms    14.203ms 1000469.800       0.000 178249093.1  75169253.2
ListKafkaConnectionsTest.CollectProto250k                        2    34.756ms   372.677us    34.257ms    43.658ms 2501158.000       0.000 445639345.2 196148938.8
ListKafkaConnectionsTest.UnorderedInsertExtract100k              5     9.337ms   277.948us     9.028ms    10.132ms  200796.000      13.560  67508290.2  50832862.5
ListKafkaConnectionsTest.UnorderedInsertExtract250k              2    23.533ms   292.763us    23.240ms    24.140ms  501969.000      31.100 168758142.4 128303692.6
ListKafkaConnectionsTest.OrderedInsert10k                       46     2.656ms   201.230us     2.422ms     3.194ms  135620.800       7.996  37673676.0  14411238.5
ListKafkaConnectionsTest.OrderedInsert100k                       5    37.106ms   565.834us    36.313ms    38.040ms 1359486.200      69.200 377839668.0 200662008.1
ListKafkaConnectionsTest.OrderedInsert250k                       2    93.537ms   303.527us    91.483ms    93.841ms 3399565.000     170.300 944830310.9 503646360.4
ListKafkaConnectionsTest.OrderedExtract10k                      28    10.217ms   199.791us     9.952ms    10.692ms  546589.314      23.414 160864658.7  54632298.4
ListKafkaConnectionsTest.OrderedExtract100k                      3   202.626ms   417.978us   196.329ms   205.021ms 6798893.067     406.667 2008655124.0 1091604280.9
ListKafkaConnectionsTest.OrderedExtract250k                      1   649.845ms     5.973ms   643.872ms   693.044ms 18294883.600    1324.000 5411307332.0 3587524143.4
ListKafkaConnectionsTest.ToProto10k                             39     5.769ms    13.281us     5.747ms     5.799ms  220072.000      23.462 125429038.8  30730835.5
ListKafkaConnectionsTest.ToProto100k                             4    61.049ms   835.368us    59.606ms    62.204ms 2200517.200     239.250 1254310351.1 323518758.4
ListKafkaConnectionsTest.ToJson10k                              22    26.352ms   105.037us    26.229ms    26.468ms  280186.000     105.727 456151590.1 140718130.8
ListKafkaConnectionsTest.ToJson100k                              3   265.074ms   469.551us   264.474ms   265.544ms 2801650.667    1055.667 4561772366.7 1412622467.7
ListKafkaConnectionsTest.UnorderedInsertProto100of100k           6     5.691ms    20.413us     5.659ms     6.281ms     204.000       0.000  61350747.5  31311410.8
ListKafkaConnectionsTest.OrderedInsertProto100of100k             5    20.093ms   239.787us    19.840ms    21.954ms  431978.400      31.120 155769734.6 110934282.3
ListKafkaConnectionsTest.E2EUnorderedProto10k                   38     7.854ms    39.901us     7.701ms     7.932ms  340210.547      26.889 149964546.3  41871378.8
ListKafkaConnectionsTest.E2EUnorderedProto100k                   4    83.534ms   233.196us    82.061ms    83.933ms 3401772.300     248.750 1500028417.0 443883801.9
ListKafkaConnectionsTest.E2EUnorderedJson10k                    21    28.894ms    61.803us    28.716ms    29.028ms  400308.114     110.476 480687347.9 154071789.9
ListKafkaConnectionsTest.E2EUnorderedJson100k                    3   290.131ms   351.367us   288.762ms   292.523ms 4002851.667    1073.000 4807471054.8 1548414865.1
ListKafkaConnectionsTest.E2EOrderedProto10k                     25    20.858ms   145.508us    20.482ms    21.205ms 1002342.928      57.856 341769348.9 111351247.2
ListKafkaConnectionsTest.E2EOrderedProto100k                     2   341.600ms     2.084ms   333.316ms   344.391ms 11357941.000     797.500 3818749405.2 1841880691.2
ListKafkaConnectionsTest.E2EOrderedJson10k                      16    41.903ms   164.827us    41.666ms    42.206ms 1062604.300     142.050 672536096.0 224024166.7
ListKafkaConnectionsTest.E2EOrderedJson100k                      2   558.536ms     5.947ms   551.798ms   566.316ms 11958270.600    1667.700 7125728506.1 2999548876.7
ListKafkaConnectionsTest.E2EUnorderedProto100of100k              5    20.173ms   125.258us    19.809ms    20.298ms 1002891.000       0.600 240834246.5 108531102.9
ListKafkaConnectionsTest.E2EOrderedProto100of100k                5    36.553ms   469.725us    35.823ms    37.999ms 1437420.840      34.320 336054884.2 199054218.1
ListKafkaConnectionsTest.E2EUnorderedJson100of100k               5    21.770ms   417.547us    20.445ms    22.661ms 1003497.520       2.120 244150566.0 117065849.3
ListKafkaConnectionsTest.E2EOrderedJson100of100k                 5    37.656ms     1.197ms    35.681ms    39.333ms 1437906.560      35.840 339330154.2 204563851.0
```
Notes:
* For collection, ordered is about 3x unordered, for insertion.
* When accounting for insert and extract, ordered is much slower (18x for 100k connections), dominated by the `sort_heap`.
* Serialization to json is 4x slower than proto (449-165) / (233-165)

Allocations (probably an underestimate due to not all fields being added) due to `make_data()` (a proxy for `connection_context::to_proto()`) are, per connection:
* 10
* 1.4KiB

Serialization to iobuf increases that further, since both the proto and the serialized form are held in memory at the same time.
Proto: 1KiB
Json 2.5KiB


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

BenPope committed:
admin: Add a scale test for ListKafkaConnections

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
admin/kafka_connections_service: Add benchmark

ListKafkaConnectionsTest.CollectProto100k                        5    13.781ms    76.531us    13.705ms    14.203ms 1000469.800       0.000 178249093.1  75169253.2
ListKafkaConnectionsTest.CollectProto250k                        2    34.756ms   372.677us    34.257ms    43.658ms 2501158.000       0.000 445639345.2 196148938.8
ListKafkaConnectionsTest.UnorderedInsertExtract100k              5     9.337ms   277.948us     9.028ms    10.132ms  200796.000      13.560  67508290.2  50832862.5
ListKafkaConnectionsTest.UnorderedInsertExtract250k              2    23.533ms   292.763us    23.240ms    24.140ms  501969.000      31.100 168758142.4 128303692.6
ListKafkaConnectionsTest.OrderedInsert10k                       46     2.656ms   201.230us     2.422ms     3.194ms  135620.800       7.996  37673676.0  14411238.5
ListKafkaConnectionsTest.OrderedInsert100k                       5    37.106ms   565.834us    36.313ms    38.040ms 1359486.200      69.200 377839668.0 200662008.1
ListKafkaConnectionsTest.OrderedInsert250k                       2    93.537ms   303.527us    91.483ms    93.841ms 3399565.000     170.300 944830310.9 503646360.4
ListKafkaConnectionsTest.OrderedExtract10k                      28    10.217ms   199.791us     9.952ms    10.692ms  546589.314      23.414 160864658.7  54632298.4
ListKafkaConnectionsTest.OrderedExtract100k                      3   202.626ms   417.978us   196.329ms   205.021ms 6798893.067     406.667 2008655124.0 1091604280.9
ListKafkaConnectionsTest.OrderedExtract250k                      1   649.845ms     5.973ms   643.872ms   693.044ms 18294883.600    1324.000 5411307332.0 3587524143.4
ListKafkaConnectionsTest.ToProto10k                             39     5.769ms    13.281us     5.747ms     5.799ms  220072.000      23.462 125429038.8  30730835.5
ListKafkaConnectionsTest.ToProto100k                             4    61.049ms   835.368us    59.606ms    62.204ms 2200517.200     239.250 1254310351.1 323518758.4
ListKafkaConnectionsTest.ToJson10k                              22    26.352ms   105.037us    26.229ms    26.468ms  280186.000     105.727 456151590.1 140718130.8
ListKafkaConnectionsTest.ToJson100k                              3   265.074ms   469.551us   264.474ms   265.544ms 2801650.667    1055.667 4561772366.7 1412622467.7
ListKafkaConnectionsTest.UnorderedInsertProto100of100k           6     5.691ms    20.413us     5.659ms     6.281ms     204.000       0.000  61350747.5  31311410.8
ListKafkaConnectionsTest.OrderedInsertProto100of100k             5    20.093ms   239.787us    19.840ms    21.954ms  431978.400      31.120 155769734.6 110934282.3
ListKafkaConnectionsTest.E2EUnorderedProto10k                   38     7.854ms    39.901us     7.701ms     7.932ms  340210.547      26.889 149964546.3  41871378.8
ListKafkaConnectionsTest.E2EUnorderedProto100k                   4    83.534ms   233.196us    82.061ms    83.933ms 3401772.300     248.750 1500028417.0 443883801.9
ListKafkaConnectionsTest.E2EUnorderedJson10k                    21    28.894ms    61.803us    28.716ms    29.028ms  400308.114     110.476 480687347.9 154071789.9
ListKafkaConnectionsTest.E2EUnorderedJson100k                    3   290.131ms   351.367us   288.762ms   292.523ms 4002851.667    1073.000 4807471054.8 1548414865.1
ListKafkaConnectionsTest.E2EOrderedProto10k                     25    20.858ms   145.508us    20.482ms    21.205ms 1002342.928      57.856 341769348.9 111351247.2
ListKafkaConnectionsTest.E2EOrderedProto100k                     2   341.600ms     2.084ms   333.316ms   344.391ms 11357941.000     797.500 3818749405.2 1841880691.2
ListKafkaConnectionsTest.E2EOrderedJson10k                      16    41.903ms   164.827us    41.666ms    42.206ms 1062604.300     142.050 672536096.0 224024166.7
ListKafkaConnectionsTest.E2EOrderedJson100k                      2   558.536ms     5.947ms   551.798ms   566.316ms 11958270.600    1667.700 7125728506.1 2999548876.7
ListKafkaConnectionsTest.E2EUnorderedProto100of100k              5    20.173ms   125.258us    19.809ms    20.298ms 1002891.000       0.600 240834246.5 108531102.9
ListKafkaConnectionsTest.E2EOrderedProto100of100k                5    36.553ms   469.725us    35.823ms    37.999ms 1437420.840      34.320 336054884.2 199054218.1
ListKafkaConnectionsTest.E2EUnorderedJson100of100k               5    21.770ms   417.547us    20.445ms    22.661ms 1003497.520       2.120 244150566.0 117065849.3
ListKafkaConnectionsTest.E2EOrderedJson100of100k                 5    37.656ms     1.197ms    35.681ms    39.333ms 1437906.560      35.840 339330154.2 204563851.0

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
admin/kafka_connections_service: Extract impl

No functional change.

Allow testing of the collectors.

Signed-off-by: Ben Pope <ben@redpanda.com>

pgellert reviewed (commented):
Looks pretty good, thanks.

Looking at the benchmarks: If I am reading this correctly, this means that in the 100k connection case, sorting is ~157ms additional latency vs unsorted; then serialization adds another ~68ms if proto or ~283ms if json. There might be some micro-optimizations we could do here, but that seems reasonable to be.

Some additional scenarios that would be nice to get benchmarks for are:
* OrderedProto and OrderedJSON numbers for the 1k and 10k use cases
* OrderedProto and OrderedJSON (or ExtractOrdered) for a scenario where there are 100k connections to search through, but the page_size limit is ~20 (rpk default) or maybe 100 or some other low number like that.

Also, it's worth noting that the real latencies will also include the latencies of the cross-core and cross-broker calls, so I would expect that the real latencies are probably somewhere in the range of ~2x from here (sorting is done first per-broker, sequentially, and then combined on the "external rpc handling" node).

BenPope commented:
> Looking at the benchmarks: If I am reading this correctly, this means that in the 100k connection case, sorting is ~157ms additional latency vs unsorted; then serialization adds another ~68ms if proto or ~283ms if json.

Yes.

> Some additional scenarios that would be nice to get benchmarks for are:
> 
> * OrderedProto and OrderedJSON numbers for the 1k and 10k use cases

Added.

> * OrderedProto and OrderedJSON (or ExtractOrdered) for a scenario where there are 100k connections to search through, but the page_size limit is ~20 (rpk default) or maybe 100 or some other low number like that.

Added, with a limit of 100.

> Also, it's worth noting that the real latencies will also include the latencies of the cross-core and cross-broker calls, so I would expect that the real latencies are probably somewhere in the range of ~2x from here (sorting is done first per-broker, sequentially, and then combined on the "external rpc handling" node).

Do you think it's worth adding something this to the benchmark?

Say 1000 connections, 8 cores, 8 machines? To run through the intermediate steps?

pgellert commented:
> Do you think it's worth adding something this to the benchmark?
> Say 1000 connections, 8 cores, 8 machines? To run through the intermediate steps?

In an ideal world, I would love to have a benchmark for `kafka_connections_service::list_kafka_connections_local` as well, as a benchmark against e2e performance regressions. But that is probably overly complex to set up, more of a nice to have, and lower priority than just having an understanding of what the expected e2e latency looks like and to see if any hard limits are necessary in rpk / broker-side.

I think we need an understanding of what the e2e latency will look like in a representative real cluster scenario like the `1000 connections, 8 cores, 8 machines` or the LRC or a loaded Tier 5 BYOC cluster, etc. I don't know if we can even have multi-node "micro-benchmarks", so I think that would have to be either manually tested or using the ducktape scale test you have in this PR.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/6395fe4c157160d022d1ffe2903c9843d710b1af..994e4dce81994d7271e071f90a25fc71d5bf2950)
Improve the benchmark
* Separate constructing the proto type as its own phase, distinct from generating the data
* Add scale benchmarks for each phase independently
* Rename and reorder benchmarks to more clearly show scaling

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--50a86b62-ecdf-479b-891d-45bc4f8a2e09-->
<!--TEST RESULTS-->
<!--7337632d-c0fc-4046-bce2-72e1c3fc8dbc-->
<!--TEST RESULTS-->
<!--3d6c7073-45e2-48c0-85ea-035a02f2ee71-->
<!--TEST RESULTS-->
<!--c217b80b-daf6-41bc-aed7-60095b5f13c2-->
<!--TEST RESULTS-->
<!--8c518c79-be32-488c-9466-b8e88772d316-->
<!--TEST RESULTS-->
<!--842a12af-8dcc-4983-85da-9cee631cf9bd-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75074">build#75074</a></summary>


| test_class                                                     | test_method                                                    | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                               |
|:---------------------------------------------------------------|:---------------------------------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalUpgradeTest                              | test_tx_control_batch_removal_with_upgrade                     | {"test_case_name": "All commits"} | integration | https://buildkite.com/redpanda/redpanda/builds/75074#019a281e-1110-4770-a0d2-3a5ac589aae3 | FLAKY         | 19/21    | upstream reliability is '93.33333333333333'. current run reliability is '90.47619047619048'. drift is 2.85714 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| src/v/redpanda/admin/tests/list_kafka_connections_rpbench_test | src/v/redpanda/admin/tests/list_kafka_connections_rpbench_test |                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/75074#019a27f8-5a80-480b-a186-9b8c5cafaa1f | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                                            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75121">build#75121</a></summary>


| test_class                 | test_method                              | test_arguments          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                      |
|:---------------------------|:-----------------------------------------|:------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest  | test_write                               | null                    | integration | https://buildkite.com/redpanda/redpanda/builds/75121#019a2b6b-09e8-404f-80ec-9ff26b29ab16 | FLAKY         | 18/21    | upstream reliability is '95.86919104991394'. current run reliability is '85.71428571428571'. drift is 10.15491 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                |
| ShardPlacementTest         | test_upgrade                             | null                    | integration | https://buildkite.com/redpanda/redpanda/builds/75121#019a2b6b-09ee-4de2-8aa4-3a3ac1903d53 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShardPlacementTest&test_method=test_upgrade                                     |
| TLSVersionTestRSA          | test_change_version                      | {"version": 1}          | integration | https://buildkite.com/redpanda/redpanda/builds/75121#019a2b67-56eb-40d1-b2c0-037aca598c1d | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSVersionTestRSA&test_method=test_change_version                               |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75121#019a2b67-56ed-4e98-a81e-e5868e35e3fd | FLAKY         | 20/21    | upstream reliability is '99.34497816593887'. current run reliability is '95.23809523809523'. drift is 4.10688 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75136">build#75136</a></summary>


| test_class                           | test_method                       | test_arguments                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                    | test_history                                                                                                                                                                         |
|:-------------------------------------|:----------------------------------|:------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AdminV2ListKafkaConnectionsScaleTest | test_list_kafka_connections_scale | {"consumers_per_group": 2500, "group_count": 4, "ordered": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75136#019a2be1-0bf7-4a39-9b6b-e94af9d9e37f | FLAKY         | 1/21     | The test was found to be new, and no failures are allowed | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsScaleTest&test_method=test_list_kafka_connections_scale |
| AdminV2ListKafkaConnectionsScaleTest | test_list_kafka_connections_scale | {"consumers_per_group": 2500, "group_count": 4, "ordered": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75136#019a2be3-a689-4b88-aa99-9ba98bf6639a | FAIL          | 0/21     | The test has failed across all retries                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsScaleTest&test_method=test_list_kafka_connections_scale |
| AdminV2ListKafkaConnectionsScaleTest | test_list_kafka_connections_scale | {"consumers_per_group": 2500, "group_count": 4, "ordered": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/75136#019a2be3-a689-4b88-aa99-9ba98bf6639a | FAIL          | 0/21     | The test has failed across all retries                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsScaleTest&test_method=test_list_kafka_connections_scale |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75306">build#75306</a></summary>


| test_class                        | test_method                                | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                               |
|:----------------------------------|:-------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests     | test_auto_prefix_trimming                  | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75306#019a34be-4e07-4016-a966-b7edeaa2ca7f | FLAKY         | 18/21    | upstream reliability is '92.90953545232273'. current run reliability is '85.71428571428571'. drift is 7.19525 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming                      |
| MountUnmountIcebergTest           | test_simple_remount                        | {"cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75306#019a34be-4e08-4839-9557-de1f95b280fd | FLAKY         | 13/21    | upstream reliability is '77.9683377308707'. current run reliability is '61.904761904761905'. drift is 16.06358 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                                  |
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "All commits"}                                            | integration | https://buildkite.com/redpanda/redpanda/builds/75306#019a34be-4e08-4839-9557-de1f95b280fd | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "Multi-segment spanning transactions"}                    | integration | https://buildkite.com/redpanda/redpanda/builds/75306#019a34bb-937a-4912-ae33-87ac29bc25ed | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| PartitionMovementTest             | test_empty                                 | {"num_to_upgrade": 0}                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75306#019a34be-4e03-4563-8311-d7ef698d43e7 | FLAKY         | 19/21    | upstream reliability is '87.4296435272045'. current run reliability is '90.47619047619048'. drift is -3.04655 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                                             |
| PartitionMovementTest             | test_static                                | {"num_to_upgrade": 0}                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75306#019a34be-4e07-4016-a966-b7edeaa2ca7f | FLAKY         | 18/21    | upstream reliability is '87.5'. current run reliability is '85.71428571428571'. drift is 1.78571 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                                            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75324">build#75324</a></summary>


| test_class                           | test_method                       | test_arguments                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                    | test_history                                                                                                                                                                         |
|:-------------------------------------|:----------------------------------|:------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AdminV2ListKafkaConnectionsScaleTest | test_list_kafka_connections_scale | {"consumers_per_group": 2500, "group_count": 4, "ordered": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75324#019a358d-ab9e-4756-b27b-73efdc3ab6fa | FLAKY         | 1/21     | The test was found to be new, and no failures are allowed | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsScaleTest&test_method=test_list_kafka_connections_scale |
| AdminV2ListKafkaConnectionsScaleTest | test_list_kafka_connections_scale | {"consumers_per_group": 2500, "group_count": 4, "ordered": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/75324#019a358d-ab9e-4756-b27b-73efdc3ab6fa | FLAKY         | 1/21     | The test was found to be new, and no failures are allowed | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsScaleTest&test_method=test_list_kafka_connections_scale |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75395">build#75395</a></summary>


| test_class                           | test_method                          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                  |
|:-------------------------------------|:-------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests                 | test_task_pausing                    | {"shuffle_leadership": false}                                                                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-6f39-43c5-a3e5-c54bb694abdb | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_task_pausing                          |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync           | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-91ad-47aa-8b76-457fa7cbf36e | FLAKY         | 15/21    | upstream reliability is '97.9633401221996'. current run reliability is '71.42857142857143'. drift is 26.53477 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync           | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-91b0-4e89-9c49-0dde70aca68a | FLAKY         | 17/21    | upstream reliability is '98.55855855855856'. current run reliability is '80.95238095238095'. drift is 17.60618 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkingReplicationTests        | test_replication_timestamps_match    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "timestamp_type": "CreateTime"}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-6f2f-4a46-a77c-7663ee44b661 | FLAKY         | 20/21    | upstream reliability is '95.2755905511811'. current run reliability is '95.23809523809523'. drift is 0.0375 and the allowed drift is set to 50. The test should PASS     | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| DataMigrationsApiTest                | test_creating_and_listing_migrations | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-91aa-4185-96b0-eee695959495 | FLAKY         | 19/21    | upstream reliability is '99.43555973659454'. current run reliability is '90.47619047619048'. drift is 8.95937 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations      |
| MountUnmountIcebergTest              | test_simple_remount                  | {"cloud_storage_type": 1}                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-6f2f-4a46-a77c-7663ee44b661 | FLAKY         | 12/21    | upstream reliability is '76.74418604651163'. current run reliability is '57.14285714285714'. drift is 19.60133 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                     |
| NodePostRestartProbeTest             | post_restart_probe_test              | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-6f34-4887-bb82-002ddebf6486 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodePostRestartProbeTest&test_method=post_restart_probe_test                |
| PartitionMovementTest                | test_static                          | {"num_to_upgrade": 0}                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-6f39-43c5-a3e5-c54bb694abdb | FLAKY         | 13/21    | upstream reliability is '87.55221386800334'. current run reliability is '61.904761904761905'. drift is 25.64745 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                               |
| PartitionMovementUpgradeTest         | test_basic_upgrade                   | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-91b3-4fb1-99c8-3ea4f6135fcc | FLAKY         | 19/21    | upstream reliability is '99.78070175438597'. current run reliability is '90.47619047619048'. drift is 9.30451 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade                 |
| SimpleEndToEndTest                   | test_relaxed_acks                    | {"write_caching": false}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75395#019a3945-6f2f-4a46-a77c-7663ee44b661 | FLAKY         | 20/21    | upstream reliability is '98.94736842105263'. current run reliability is '95.23809523809523'. drift is 3.70927 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SimpleEndToEndTest&test_method=test_relaxed_acks                            |
</details>
<!--/body-->

pgellert commented on src/v/redpanda/admin/tests/list_kafka_connections_bench.cc:
```diff
@@ -0,0 +1,325 @@
+                         const proto::admin::kafka_connection& a,
+                         const proto::admin::kafka_connection& b) {
+                return a.get_uid() < b.get_uid();
+            };
```
nit: this could be made even more realistic:
```
auto cfg
  = admin::make_ordering_config<proto::admin::kafka_connection>(
    "uid");
auto cmp = admin::sort_order::parse(std::move(cfg));
```

BenPope replied:
Good call, Ordered Insert and Extract are now 40% slower.

pgellert commented on src/v/redpanda/admin/tests/list_kafka_connections_bench.cc:
```diff
@@ -0,0 +1,325 @@
+}
+PERF_TEST_F(ListKafkaConnectionsTest, UnorderedExtract100k) {
+    return run_test_impl(false, false, false, true, std::nullopt, k100, k100);
+}
```
The benchmarking guidance from @travisdowns and @StephanDollberg says that each run should be at least in the ~1ms range:
> So you do not need a very large number of internal iterations to hit a long overall runtime (to reduce noise). Rather, in order to play nice with this approach, a single invocation of the **benchmark methods should run a relatively short period of time (but not too short, per above), e.g., in the 1 ms to 100 ms range.** This leads to a benchmark that responds to changes in the --duration argument, which is important for the _test variations which set this value to 0. 

Some of these O(1) runs are <1ms and so I think they could become overly noisy. I think we could either disable them with a flag, drop them or add more iterations to them as a `PERF_TEST_CN`. I am fine with either option, probably disabling with a hardcoded constant flag is easiest.

BenPope replied:
I dropped the 1k runs.

pgellert reviewed (commented):
Looks great!

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75121-->
<!--1cd8bc15-474d-4e69-a0f0-a3089dab1ce3-->
<!--title-->
## Retry command for [Build#75121](https://buildkite.com/redpanda/redpanda/builds/75121)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tls_version_test.py::TLSVersionTestRSA.test_change_version@{"version":1}
```
<!--/body-->

BenPope commented:
/ci-repeat 1
tests/rptest/tests/tls_version_test.py::TLSVersionTestRSA.test_change_version@{"version":1}
tests/rptest/scale_tests/list_kafka_connections_scale_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75136-->
<!--6f1d7187-78a7-4b28-b12e-60b1e2d2eb32-->
<!--2ffc258f-cf27-48a6-aa4b-7d7fef99dca3-->
<!--title-->
## Retry command for [Build#75136](https://buildkite.com/redpanda/redpanda/builds/75136)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/list_kafka_connections_scale_test.py::AdminV2ListKafkaConnectionsScaleTest.test_list_kafka_connections_scale@{"consumers_per_group":2500,"group_count":4,"ordered":false}
tests/rptest/scale_tests/list_kafka_connections_scale_test.py::AdminV2ListKafkaConnectionsScaleTest.test_list_kafka_connections_scale@{"consumers_per_group":2500,"group_count":4,"ordered":true}
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75306-->
<!--ef741f2d-5c64-4286-a873-14198a578748-->
<!--title-->
## Retry command for [Build#75306](https://buildkite.com/redpanda/redpanda/builds/75306)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/log_compaction_test.py::LogCompactionTxRemovalUpgradeTest.test_tx_control_batch_removal_with_upgrade@{"test_case_name":"Multi-segment spanning transactions"}
```
<!--/body-->

BenPope commented:
/ci-repeat 1
tests/rptest/tests/log_compaction_test.py::LogCompactionTxRemovalUpgradeTest.test_tx_control_batch_removal_with_upgrade@{"test_case_name":"Multi-segment spanning transactions"}
tests/rptest/scale_tests/list_kafka_connections_scale_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75324-->
<!--eaf09404-0e2d-4cdf-95b0-3f8cdcd36177-->
<!--title-->
## Retry command for [Build#75324](https://buildkite.com/redpanda/redpanda/builds/75324)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/list_kafka_connections_scale_test.py::AdminV2ListKafkaConnectionsScaleTest.test_list_kafka_connections_scale@{"consumers_per_group":2500,"group_count":4,"ordered":false}
tests/rptest/scale_tests/list_kafka_connections_scale_test.py::AdminV2ListKafkaConnectionsScaleTest.test_list_kafka_connections_scale@{"consumers_per_group":2500,"group_count":4,"ordered":true}
```
<!--/body-->

BenPope committed:
dt/rpk: Support cluster connections list

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/641e08eb6b6bb16f5cbb3a83efe841001b09f1d5..f687d0293e0ec9fda463749d3d93a59cc2277283)
* Rebased

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f687d0293e0ec9fda463749d3d93a59cc2277283..9478c52aa014934337c8fd78c41d00d62b44fd07)
* Add RPK support for `cluster connections list`
* Use rpk for the scale test (locally this runs in 0.62s instead of 1.4s)

BenPope commented:
/ci-repeat 1
dt-repeat=10
release
tests/rptest/scale_tests/list_kafka_connections_scale_test.py
